[editor][client] Enable Telemetry based on User Config settings

Ankush-lastmile opened the pull request:
[editor][client] Enable Telemetry based on User Config settings




This diff builds on top of @rholinshead's starting point for telemetry diff. Most of that looked good to me so I didn't touch it beside @rossdanlm 's .


- Building off of #869, if `allow_usage_data_sharing` is set to True, initialize Datadog Browser logging with a session ID.
- Disabled telemetry in dev mode as per @rholinshead 's comment that the hot reload spams the logs with something like "datadog logging already initialized"
- Moved the initialization logic away from`index.tsx` into `editor.tsx` so that it can be configurable.


## Testplan:

1. yarn build
2. run in "prod" mode
3. Edit AIconfig
4. Hit Save button -> validate datadog log sent

repeat with aiconfig.rc set with false logging

|<img width="963" alt="Screenshot 2024-01-12 at 7 24 33 PM" src="https://github.com/lastmile-ai/aiconfig/assets/141073967/c2f979df-327e-40c4-9f06-034531437a65">  | <img width="1455" alt="Screenshot 2024-01-12 at 7 23 51 PM" src="https://github.com/lastmile-ai/aiconfig/assets/141073967/edb022b5-4abd-4ddc-b9bb-97713d32e5dc"> |
| ------------- | ------------- |
|
<img width="604" alt="Screenshot 2024-01-12 at 7 26 35 PM" src="https://github.com/lastmile-ai/aiconfig/assets/141073967/4a2e49a3-061e-404c-9681-13a15aca8e9c">  |  -> No logs |




I tried taking a video but datadog doesn't immediately update with logs which made the video too long to upload

---
Stack created with [Sapling](https://sapling-scm.com). Best reviewed with [ReviewStack](https://reviewstack.dev/lastmile-ai/aiconfig/pull/899).
* __->__ #899
* #864

Ankush-lastmile commented on python/src/aiconfig/editor/client/src/LocalEditor.tsx:
```diff
@@ -31,6 +26,31 @@ export default function Editor() {
+
+      datadogLogs.setGlobalContext({ session_id: sessionID });
+    }
+  }, []);
```
setGlobalContext will add this as an attribute to every log sent

Ankush-lastmile replied:
also should clientToken be a constant somewhere? was unsure if that should be the case

rossdanlm commented:
> This diff builds on top of @rholinshead's starting point for telemetry diff.

Can you link diff?

rossdanlm commented on python/src/aiconfig/editor/client/src/LocalEditor.tsx:
```diff
@@ -29,6 +26,31 @@ export default function Editor() {
+
+    if (enableTelemetry) {
+      datadogLogs.init({
+        clientToken: "pub356987caf022337989e492681d1944a8",
```
What's this?

Ankush-lastmile replied:
datadog browser logging uses this token for authentication

Ankush-lastmile replied:
its generated in datadog (done in the datadog ui probably by ryan)

rossdanlm commented on python/src/aiconfig/editor/client/src/LocalEditor.tsx:
```diff
@@ -29,6 +26,31 @@ export default function Editor() {
+      datadogLogs.init({
+        clientToken: "pub356987caf022337989e492681d1944a8",
+        env: process.env.NODE_ENV ?? "development",
+        service: "aiconfig-editor",
+        site: "us5.datadoghq.com",
+        forwardErrorsToLogs: true,
+        sessionSampleRate: 100,
```
In diff summary, can you link to website for API docs so we can read more on what these mean?

Ankush-lastmile replied:
whoops forgot to link docs. Generally, datadog docs are easily googleable and easily navigate able. Can just google "datadog <search term>" 

"datadog browser logging" ->  https://docs.datadoghq.com/logs/log_collection/javascript/

rossdanlm commented on python/src/aiconfig/editor/client/src/utils/logging.ts:
```diff
@@ -0,0 +1,6 @@
+
+
+export function generateUniqueSessionId(): string {
+    return uuid.v4();
```
Oh nice, just like what fb users

rossdanlm commented on python/src/aiconfig/editor/client/src/utils/logging.ts:
```diff
@@ -0,0 +1,6 @@
+
+export function generateUniqueSessionId(): string {
+    return uuid.v4();
+}
```
nit: add new line

rossdanlm commented on python/src/aiconfig/editor/client/src/utils/logging.ts:
```diff
@@ -0,0 +1,6 @@
+import * as uuid from "uuid";
+
```
nit: is it normal to have two empty lines or just 1?

Ankush-lastmile replied:
removed custom id

rossdanlm commented on python/src/aiconfig/editor/client/src/shared/types.ts:
```diff
@@ -61,3 +61,6 @@ export function aiConfigToClientConfig(aiconfig: AIConfig): ClientAIConfig {
   };
 }
+
+export type LogEvent = "ADD_PROMPT" | "SAVE_BUTTON_CLICKED";
```
Nice, good start

rossdanlm commented on python/src/aiconfig/editor/client/src/LocalEditor.tsx:
```diff
@@ -174,6 +196,14 @@ export default function Editor() {
+  const logEvent = useCallback((event: LogEvent, data?: LogEventData) => {
+    try {
+      datadogLogs.logger.info(event, data);
+    } catch (e) {
+      // Ignore logger errors for now
+    }
```
Does this get handled automatically if we remove try-catch by `forwardErrorsToLogs` field?

Ankush-lastmile replied:
Two seperate things.

This try catch ignores any errors by datadog logging module.

`forwardErrorsToLogs` forwards any `console.warn` to logs

rossdanlm commented on python/src/aiconfig/editor/client/src/LocalEditor.tsx:
```diff
@@ -29,6 +26,31 @@ export default function Editor() {
     loadConfig();
   }, [loadConfig]);
 
+  const enableTelemetryIfAllowed = useCallback(async () => {
```
nit: `enable` implies that we're turning on a boolean flag. Let's change this to `setupTelemetryIfAllowed`

Ankush-lastmile replied:
updated, thanks for the suggestion

rossdanlm commented on python/src/aiconfig/editor/client/src/shared/types.ts:
```diff
@@ -61,3 +61,6 @@ export function aiConfigToClientConfig(aiconfig: AIConfig): ClientAIConfig {
 }
+
+export type LogEvent = "ADD_PROMPT" | "SAVE_BUTTON_CLICKED";
+export type LogEventData = JSONObject;
```
not a blocker, but let's add an issue to schematize the logging payload

Ankush-lastmile replied:
added a todo

Ankush-lastmile replied:
are we tracking editor issues with github issues?

rholinshead replied:
We should, TODOs were mainly from when it was pre-release

Ankush-lastmile replied:
sounds good, created #915

rossdanlm commented on python/src/aiconfig/editor/client/src/LocalEditor.tsx:
```diff
@@ -174,6 +196,14 @@ export default function Editor() {
 
+  const logEvent = useCallback((event: LogEvent, data?: LogEventData) => {
+    try {
+      datadogLogs.logger.info(event, data);
```
So if telemetry is false, this is going to error? Instead of try catch I think just checking for non-null and no-op is better? We can stil wrap the actual `datadogLogs.logger.info(event, data)` call in a try-catch

Ankush-lastmile replied:
I manually tested this and this doesn't error, it just becomes a no-op

rossdanlm commented on python/src/aiconfig/editor/client/src/components/AIConfigContext.tsx:
```diff
@@ -1,12 +1,13 @@
  */
 const AIConfigContext = createContext<{
   getState: () => ClientAIConfig;
+  logEvent?: (event: LogEvent, data?: LogEventData) => void;
```
Why do we need to pass this in? Does it get called anywhere in this PR?

Ankush-lastmile replied:
I think this was answered by yourself so resolving

rossdanlm commented on python/src/aiconfig/editor/client/src/components/AIConfigEditor.tsx:
```diff
@@ -771,8 +777,9 @@ export default function EditorContainer({
   const contextValue = useMemo(
     () => ({
       getState,
+      logEvent,
```
Oh I see, it's passed into here from the editor wrapper

rossdanlm commented on python/src/aiconfig/editor/client/src/components/AIConfigEditor.tsx:
```diff
@@ -171,6 +174,8 @@ export default function EditorContainer({
   const stateRef = useRef(aiconfigState);
   stateRef.current = aiconfigState;
 
+  const logEvent = callbacks.logEvent;
```
nit: a bit confusing since this isn't the `logEvent`, it's a `logEventCallback`, can we rename it to that (as well as prop name within `AIConfigContext`)?

Ankush-lastmile replied:
updated per your suggestion

rossdanlm commented on python/src/aiconfig/editor/client/src/components/AIConfigEditor.tsx:
```diff
@@ -878,7 +885,10 @@ export default function EditorContainer({
-                onClick={onSave}
+                onClick={() => {
+                  onSave();
+                  logEvent?.("SAVE_BUTTON_CLICKED");
```
Edit: Oh nm I see that the `data` field is optional: `data?: LogEventData`. Makes sense

** Old **
Do we need to pass in empty event data here?

```
logEvent?.("SAVE_BUTTON_CLICKED", {})
```

rossdanlm replied:
Ok so in general how FB did logging, we'd have a field for UI component/surface, and then another field for EVENT (generated, clicked, etc).

But yea not a blocker, just FYI, this is good for now

Ankush-lastmile replied:
thanks for the insight!

rossdanlm reviewed (approved):
Generally lgtm, just some comments and nits

rholinshead commented on python/src/aiconfig/editor/client/src/LocalEditor.tsx:
```diff
@@ -31,6 +26,31 @@ export default function Editor() {
   }, [loadConfig]);
 
+  const enableTelemetryIfAllowed = useCallback(async () => {
+    const res = await ufetch.post(ROUTE_TABLE.LOAD, {}); // change this to the endpoint used to get the config
```
I think we probably will want to skip logging in dev (should check before sending the request) now that it's being initialized within the component mount -- just looked and it tries to initialize a ton, seems to be from hot reloading:

![Screenshot 2024-01-11 at 10 43 17 PM](https://github.com/lastmile-ai/aiconfig/assets/5060851/64eebd94-ca55-452e-a47e-fe3d00a1a95c)

Ankush-lastmile replied:
updated to only initialize when in "prod" mode as per your suggestion

Ankush-lastmile commented:
- removed custom sessionID because @rholinshead pointed out that datadog already passes those in by default.
- added in check for aiconfigrc
- addressed some nits

Ankush-lastmile commented:
> Can you link diff?

Its the diff below this one, sapling created a new pr

rholinshead commented on python/src/aiconfig/editor/client/src/LocalEditor.tsx:
```diff
@@ -29,6 +25,29 @@ export default function Editor() {
   }, [loadConfig]);
 
+  const setupTelemetryIfAllowed = useCallback(async () => {
+    const res = await ufetch.get(ROUTE_TABLE.GET_AICONFIGRC, {}); // change this to the endpoint used to get the config
```
```suggestion
    const res = await ufetch.get(ROUTE_TABLE.GET_AICONFIGRC, {});
```
can remote comment now. nit, don't think {} is needed, either

Ankush-lastmile replied:
removed comment

rholinshead commented on python/src/aiconfig/editor/client/src/LocalEditor.tsx:
```diff
@@ -29,6 +25,29 @@ export default function Editor() {
+    const res = await ufetch.get(ROUTE_TABLE.GET_AICONFIGRC, {}); // change this to the endpoint used to get the config
+
+    const enableTelemetry = res.allow_usage_data_sharing;
+    const isDev = (process.env.NODE_ENV ?? "development") === "development";
```
Let's do the dev check at the top so we don't need to make a request unnecessarily

Ankush-lastmile replied:
Checked for isDev before making the request, and if so return early.

Manually tested the flows and verified telemetry is only sent in prod and not dev

rholinshead commented on python/src/aiconfig/editor/client/src/utils/api.ts:
```diff
@@ -14,6 +14,7 @@ export const ROUTE_TABLE = {
   CANCEL: urlJoin(API_ENDPOINT, "/cancel"),
   CLEAR_OUTPUTS: urlJoin(API_ENDPOINT, "/clear_outputs"),
   DELETE_PROMPT: urlJoin(API_ENDPOINT, "/delete_prompt"),
+  GET_AICONFIGRC: urlJoin(API_ENDPOINT, "/get_aiconfigrc"),
```
What does rc mean?

Ankush-lastmile replied:
fairly certain that `rc` stands for run commands. Most people will have multiple of these on their system already, you can see which rc files you have on your own computer by running this in terminal `ls -a $HOME | grep rc`.

To define if telemetry is to be disabled, user will need to define their own `.aiconfigrc`. Will need to add this to docs as well

rholinshead reviewed (approved):
LGTM, but let's do the dev check & return before sending the request unnecessarily

Ankush-lastmile commented:
- moved isDev check to occur before making the request to get the aiconfigrc data. This way the request is only made if needed
- - Manually tested the flows and verified telemetry is only sent in prod and not dev
